src/catalog/src/durable/upgrade/v58_to_v59.rs

```diff
@@ -0,0 +1,187 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use crate::durable::upgrade::MigrationAction;
+use crate::durable::upgrade::{objects_v58 as v58, objects_v59 as v59};
+use mz_proto::wire_compatible;
+use mz_proto::wire_compatible::WireCompatible;
+
+wire_compatible!(v58::ClusterReplicaKey with v59::ClusterReplicaKey);
+wire_compatible!(v58::ClusterId  with v59::ClusterId);
+wire_compatible!(v58::RoleId  with v59::RoleId);
+wire_compatible!(v58::ReplicaLogging with v59::ReplicaLogging);
+
+/// In v59, we add a pending to cluster replica ManagedLocation
+pub fn upgrade(
+    snapshot: Vec<v58::StateUpdateKind>,
+) -> Vec<MigrationAction<v58::StateUpdateKind, v59::StateUpdateKind>> {
+    snapshot
+        .into_iter()
+        .filter_map(|update| {
+            let v58::state_update_kind::Kind::ClusterReplica(
+                v58::state_update_kind::ClusterReplica { key, value },
+            ) = update.kind.as_ref().expect("missing field")
+            else {
+                return None;
+            };
+            let (logging, managed_config) = match value {
```

Comment:
I believe you can use `value.and_then(|v| v.config.as_ref()).and_then(...)` instead of making a deep stack.